### PR TITLE
refactor(connlib): use `phoenix-channel` crate for clients

### LINKED ISSUE #2158: Combine `PhoenixChannel` implementations
- [ ] Combine the PhoenixChannel implementations between clients and relay
- [ ] Consider accepting a `String` for the `url` param in the `new` call along with a list of parameters to be added to the query string. This would allow us to pass `Secret`s all the way to _just before_ the connecting library

### PR DESCRIPTION
**Pull request by thomaseizinger**
Depends-On: #4048.
Depends-On: #4015.

Resolves: #2158.

### CODE REVIEW

File: rust/phoenix-channel/src/lib.rs
Comment (thomaseizinger) [inline rust/phoenix-channel/src/lib.rs:1]: We should extract these changes into a separate PR once we have things sorted out in here.

File: rust/phoenix-channel/src/lib.rs
Comment (thomaseizinger) [inline rust/phoenix-channel/src/lib.rs]: @conectado I don't think I understand the structure here fully. `Offline` seems to be sent for offline resources? Why are `Offline` and `Disabled` out here and then `Reason` has `KnownError` which are also variants?

Can we unify the error reporting here to a common structure?

cc @AndrewDryga

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/eventloop.rs]: I never fully understood what we are doing with the `reference` here as part of `request_connection`. It is my understanding the from the perspective of Phoenix LiveView, `reference` is a unique identifier for a request. Thus, in the `phoenix-channel` implementation, this is managed by the library and sending a request gives you an `OutboundRequestId` that you can use to match responses.

For `request_connection`, we seem to be using the reference as some kind of counter? Why do we need this? Can we just remove that?

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/eventloop.rs]: @conectado Behind this `..` is hiding a `reference` parameter. What is the purpose of passing out a reference here? `phoenix-channel` does not support setting the reference because it generates an `OutboundRequestId` for you that uniquely identifies the request.

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/eventloop.rs]: Similarly, I don't quite understand what is happening here. We seem to shove different data types into the `reference` parameter. I would like to change that.

If we need to track additional data we should have a `HashMap` of `OutboundRequestId` to whatever data we need to have from the request. (We could also a feature to `phoenix-channel` to track the request and re-emit it upon response / error).

File: rust/phoenix-channel/src/lib.rs
Comment (thomaseizinger) [reply rust/phoenix-channel/src/lib.rs]: Also, `Offline` seems to be a _domain-specific_ error (only sent as a response for certain messages) so I think it should actually go into `R` type parameter. IMO, the `phoenix-channel` library shouldn't know about domain-specific errors from particular messages.

File: rust/phoenix-channel/src/lib.rs
Comment (thomaseizinger) [inline rust/phoenix-channel/src/lib.rs]: @AndrewDryga The format here seems to differ from the other error responses. Is it possible to align that?

I think the same applies to the "offline" error.

File: rust/phoenix-channel/src/lib.rs
Comment (AndrewDryga) [reply rust/phoenix-channel/src/lib.rs]: `offline` is an error that is returned when you try to access a resource that doesn't have online relays/gateways,
`disabled` is an error that can be returned only when you try to create a log sync and this feature is disabled. So they are unified, but you probably need to expect them to differ depending on which message was sent.

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (AndrewDryga) [reply rust/connlib/clients/shared/src/eventloop.rs]: If we talk about `ref` that is pushed by `request_connection` message and then sent back in `connection_ready` then it's our own implementation and we need it to track for which connection the reply is. The protocol has replies for the server but doesn't have them for the clients, so we built them ourselves.

BTW `Phoenix LiveView` is just for front-end, this is all Phoenix PubSub.

File: rust/phoenix-channel/src/lib.rs
Comment (AndrewDryga) [reply rust/phoenix-channel/src/lib.rs]: This is standard error response by Phoenix

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (conectado) [reply rust/connlib/clients/shared/src/eventloop.rs]: This is actually a long time left over, we no longer keep track of the resource id on the reference.

The only time we really use it, that I remember, is when we send a resource request to make sure that we only use the response for the latest one we made since we could've made many of them and the state on the portal could change if there's a hold up somewhere.

File: rust/phoenix-channel/src/lib.rs
Comment (conectado) [reply rust/phoenix-channel/src/lib.rs]: phoenix-channel errors are not very consistent, and sometimes you get a reason, sometimes you get an outside error. The way I have organized the types around it is how I found best to react to them and design strong-typing around them, the format was found heuristically since there's no spec(or the javascript client is the spec?).

Feel free to change it, I have wanted to clean this up for a long time.

Comment (AndrewDryga) [reply rust/phoenix-channel/src/lib.rs]: JS client is the spec :)

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (AndrewDryga) [reply rust/connlib/clients/shared/src/eventloop.rs]: To implement protocol properly its better to always do something like promises - always track the response id, because protocol is asynchronous. But it up to you guys to decide how you need to use it

File: rust/phoenix-channel/src/lib.rs
Comment (thomaseizinger) [reply rust/phoenix-channel/src/lib.rs]: > `offline` is an error that is returned when you try to access a resource that doesn't have online relays/gateways, `disabled` is an error that can be returned only when you try to create a log sync and this feature is disabled. So they are unified, but you probably need to expect them to differ depending on which message was sent.

Would it be possible to send `offline` and `disabled` also as a nested struct with a `reason`?

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/eventloop.rs]: > This is actually a long time left over, we no longer keep track of the resource id on the reference.

Okay, I'll remove that on `main` then to keep this PR smaller.

File: rust/phoenix-channel/src/lib.rs
Comment (thomaseizinger) [reply rust/phoenix-channel/src/lib.rs]: Does that mean the other format with `reason` is non-standard? Can we use the standard format everywhere?

File: rust/phoenix-channel/src/lib.rs
Comment (AndrewDryga) [reply rust/phoenix-channel/src/lib.rs]: How you want them to look? Do we need extra structure just to wrap an enum?

Comment (thomaseizinger) [reply rust/phoenix-channel/src/lib.rs]: It would be nice if errors would always look the same. Currently, we have the following structures (see the difference in `payload`):

### Payload with `status` and `response`
```json
{"event":"phx_reply","ref":1,"topic":"client","payload":{"status":"error","response":"disabled"}}
```

### Payload with `reason`
```json
{"event": "disconnect", "ref": null, "topic": "client", "payload": { "reason": "token_expired" }}
```

### Payload with `status `and `response` with nested `reason`
```json
{"event": "phx_reply", "ref": null, "topic": "client", "payload": { "status":"error", "response":{ "reason":"unmatched topic" }}}
```

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/eventloop.rs]: Summary from a discussion on how to handle this:

- Replace the interval with a debounce, only generate a new connection intent every 2 seconds in case the client retries.
- Don't pass a reference out with the `ConnectionIntent`, instead, generate an ID before sending it via the phoenix-channel and remember which resource ID this was for.
- When receiving the response, check if the response is the highest (i.e. last) intent we sent and only pass it on then.

Do all of the above in a separate PR on main to make integration here easier.

File: rust/phoenix-channel/src/lib.rs
Comment (thomaseizinger) [reply rust/phoenix-channel/src/lib.rs]: Aligning the `disconnect` event isn't as critical, that can stay separate.

But aligning the format across all `phx_reply` messages would be nice!

Comment (AndrewDryga) [reply rust/phoenix-channel/src/lib.rs]: This one is our application error:
```
{"event":"phx_reply","ref":1,"topic":"client","payload":{"status":"error","response":"disabled"}}
```

This one is not generated by us (it only happens when a WS process connected to the client crashes) but we can make other errors look somewhat similar structure-wise:
```
{"event": "phx_reply", "ref": null, "topic": "client", "payload": { "status":"error", "response":{ "reason":"unmatched topic" }}}
```
but this one is harder because it's not a reply, so `event` will be different even if we keep the structure similar:
```
{"event": "disconnect", "ref": null, "topic": "client", "payload": { "reason": "token_expired" }}
```
I think for `event=disconnect` you should just ignore the payload, not even sure what I would send in `payload.status` here as it is not an error and `payload.response` doesn't make logical sense too as it is not a reply.

Comment (AndrewDryga) [reply rust/phoenix-channel/src/lib.rs]: I've opened a PR to make general error replies look like this:
`{"event": "phx_reply", "ref": null, "topic": "client", "payload": { "status":"error", "response":{ "reason":"unmatched topic" }}}`
but other events will have their own unique structure (like `disconnect` does).

**Comment by thomaseizinger** [comment]
For this PR to be green, I needs the linked PRs I extracted out of it. Will leave as draft until then.

**Comment by thomaseizinger** [comment]
@conectado It would be nice to still get a review for this, once it is rebased on top main with the linked PRs merged, the tests should be green :)

(At least they were green before I extracted them all.)

**Comment by thomaseizinger** [comment]
@conectado Now rebased and ready for review.

File: rust/connlib/clients/shared/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/lib.rs:123]: This is still a bit ugly but I didn't want to make any further changes as part of this refactoring.

File: rust/connlib/clients/shared/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/lib.rs:1]: @conectado I think all the functionality in here is equivalent to what we currently have in regards to panic handling etc. Let me know if you find something that I've missed.

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (conectado) [inline rust/connlib/clients/shared/src/eventloop.rs:35]: I see the conflicts here, hah

File: rust/connlib/clients/shared/src/lib.rs
Comment (conectado) [inline rust/connlib/clients/shared/src/lib.rs:92]: Should we create a single-threaded runtime? There's no point now to having a multi-threaded one

File: rust/connlib/clients/shared/src/lib.rs
Comment (conectado) [inline rust/connlib/clients/shared/src/lib.rs]: With these changes `on_disconnect` is always called with `None` for control channel errors? I think some clients need visibility into the http errors

**Review by conectado — COMMENTED** [review]
Looks great! just left a comment that needs to be clarified before merging

File: rust/connlib/clients/shared/src/lib.rs
Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs:92]: I wanted to keep the changes as minimal as possible but yes, we should make a single-threaded on in the future.

File: rust/connlib/clients/shared/src/lib.rs
Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs]: Yeah? The reconnection happens within `PhoenixChannel`. I'll double check if any of the clients need it. Problem is that this is now an `anyhow::Error` which I can't pass on to the callbacks.

Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs]: I just checked the clients and all of them just log this error.

Comment (conectado) [reply rust/connlib/clients/shared/src/lib.rs]: The problem is that the behavior changes if we pass an error or we don't, if we do pass an error the token is cleared.

I figure this is not the best behavior for our current architecture but a quick and dirty solution is passing an empty string.

@jamilbk  is there really a point right now where we want to call `onDisconnect` and not clear the token? 

If we don't want to clear the token we can just handle it internally.

Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs]: > The problem is that the behavior changes if we pass an error or we don't, if we do pass an error the token is cleared.

Ah okay, I didn't know that but I've now found this functionality in the Android client.

We can also remove the usage of `anyhow` from the `Eventloop` for now and actually pass the error on, that is probably a cleaner change. It is still a behaviour change because we always have an error so we will always clear the token now.

Comment (jamilbk) [reply rust/connlib/clients/shared/src/lib.rs]: > @jamilbk is there really a point right now where we want to call onDisconnect and not clear the token?

IIRC, I believe we call `onDisconnect(null)` to signal to the app that its call to `Session.disconnect()` completed successfully. At least that's how the Android and Apple behave.

So `onDisconnect(error)` is called for an unexpected disconnect (which today is only the 401 right?) and `onDisconnect(null)` is called for an expected disconnect.

Maybe it more clear to just name these explicitly:

- `onTokenExpired()`
- `onDisconnectCompleted()`

And not have to pass an error string at all because we currently don't handle the error case, right?

If the tunnel crashes or something, the apps have cleanup logic (e.g. `deinit()`) that perform any app-side cleanup. And the error will be logged in connlib's logfile anyway.

Comment (jamilbk) [reply rust/connlib/clients/shared/src/lib.rs]: https://github.com/firezone/firezone/issues/3993

Comment (jamilbk) [reply rust/connlib/clients/shared/src/lib.rs]: > It is still a behaviour change because we always have an error so we will always clear the token now.

This will be annoying to users because anytime they quit the client or stop the tunnel (i.e. `Session.disconnect`) they'll have to sign in again.

Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs]: > > It is still a behaviour change because we always have an error so we will always clear the token now.
> 
> This will be annoying to users because anytime they quit the client or stop the tunnel (i.e. `Session.disconnect`) they'll have to sign in again.

I thougt that is the same anyway? How can I stop the tunnel without signing out? At least on Android, I have to sign in every time.

Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs]: > IIRC, I believe we call `onDisconnect(null)` to signal to the app that its call to `Session.disconnect()` completed successfully

That seems pointless to me. How can disconnect not be successful? We just de-allocate memory and close file descriptors 🤷‍♂️

IMO, the clients should be able to assume that the tunnel is instantly gone as soon as disconnect returns.

Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs]: > which today is only the 401 right?

No, could be any permanent error like the TUN device being broken or non-retryable errors in the portal connection (anything 4xx I think).

Comment (jamilbk) [reply rust/connlib/clients/shared/src/lib.rs]: > I thougt that is the same anyway? How can I stop the tunnel without signing out?

Reboot with Firezone running, Select "Disconnect and Quit" from the menubar, toggle the VPN connection off in Settings, etc.

>  At least on Android, I have to sign in every time.

If you press "sign out" then we explicitly sign you out. But we need the token to stick around until it's either deemed invalid or cleared by the user because for Always On vpn to work on Android across reboots it needs this.

This was tested yesterday by a customer, so I'm surprised it's getting cleared for you -- how are you closing the tunnel?

Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs]: Right, that does work yes! I didn't even think of toggling it that way.

Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs]: > > which today is only the 401 right?
> 
> No, could be any permanent error like the TUN device being broken or non-retryable errors in the portal connection (anything 4xx I think).

Actually that is not true. Any "fatal" error in connlib currently just gets logged! That is probably something we should fix eventually but for now, I can replicate that behaviour here.

**Comment by thomaseizinger** [comment]
The last commit does not fail the eventloop on tunnel errors. This is also what we currently do. I don't think it is correct but it is what we currently do and I see it out of scope of this PR to change that.

In https://github.com/firezone/firezone/pull/4015, I am refactoring the `phoenix-channel` to only return fatal errors. Once that is merged, we can continue here to only invalidate the client's token upon authentication errors.

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/eventloop.rs:83]: I don't think this is good but this is what we currently do. I'll open an issue about it to discuss what we actually want to happen here.

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/eventloop.rs:81]: Tunnel errors **do not** end up calling `onDisconnect`. They are just logged here. This is also what we currently do on `main`. In the future, I'd like to make a type-safe error API where only very specific errors can be returned from `poll_next_event` and those will then always call `onDisconnect` to actually shut things down.

File: rust/connlib/clients/shared/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/lib.rs]: Failing to make the URL calls `on_disconnect`. This will go away as part of https://github.com/firezone/firezone/pull/4048.

File: rust/connlib/clients/shared/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/lib.rs:166]: Failing to make the tunnel calls `on_disconnect`. The only errors here are not being able to bind to the sockets and errors returned from callbacks. The former is definitely a hard error, the latter will hopefully go away at some point as we get rid of the callbacks.

File: rust/connlib/clients/shared/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/lib.rs:188]: The eventloop only ever returns fatal errors from the phoenix-channel. Those are being refactored in https://github.com/firezone/firezone/pull/4015 to really only be fatal errors.

File: rust/connlib/clients/shared/src/lib.rs
Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs:166]: I reckon we should make that one a synchronous error and directly fail the `connect` function if this goes wrong.

Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs:166]: > I reckon we should make that one a synchronous error and directly fail the `connect` function if this goes wrong.

I'll do that in a follow-up.

File: rust/connlib/clients/shared/src/lib.rs
Comment (conectado) [inline rust/connlib/clients/shared/src/lib.rs:185]: can this be changed to:
```suggestion
        Ok(_) => {},
```

**Review by conectado — APPROVED** [review]
LGTM!

File: rust/connlib/clients/shared/src/eventloop.rs
Comment (conectado) [inline rust/connlib/clients/shared/src/eventloop.rs:342]: To solve conflicts I commented out this, which was the change on control on main

File: rust/connlib/clients/shared/src/lib.rs
Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs:185]: It can but then we are not asserting that `Ok` is never actually hit and a refactoring of the return value would silently drop it!

**Pull request merged**
